pkg/controller/job/job_controller.go

```diff
@@ -702,22 +729,26 @@ func (jm *Controller) manageJob(job *batch.Job, activePods []*v1.Pod, succeeded
 		return 0, nil
 	}
 
-	rmAtLeast := active - parallelism
-	if rmAtLeast < 0 {
-		rmAtLeast = 0
-	}
-	podsToDelete := activePodsForRemoval(job, activePods, int(rmAtLeast))
-	if len(podsToDelete) > 0 {
+	if utilfeature.DefaultFeatureGate.Enabled(features.SuspendJob) && job.Spec.Suspend {
+		klog.V(4).InfoS("Deleting all active pods in suspended job", "job", klog.KObj(job), "active", active)
+		podsToDelete := activePodsForRemoval(job, activePods, int(active))
 		jm.expectations.ExpectDeletions(jobKey, len(podsToDelete))
-		klog.V(4).InfoS("Too many pods running for job", "job", klog.KObj(job), "deleted", len(podsToDelete), "target", parallelism)
 		removed, err := jm.deleteJobPods(job, jobKey, podsToDelete)
 		active -= removed
 		if err != nil {
 			return active, err
 		}
-	}
-
-	if active < parallelism {
+	} else if active > parallelism {
```

Comment:
in the new logic, we check for removals even if `active <= parallelism`. This is because there could be more than one pod with the same index, and we want to remove one of them.

I'll add a test for that.

Reply:
Created #99789

Reply:
I thought this wouldn't pass my test now

Reply:
Correct, fixed